lldb/source/Plugins/TypeSystem/Swift/TypeSystemSwiftTypeRef.cpp

```diff
@@ -157,6 +158,244 @@ TypeSystemSwiftTypeRef::CanonicalizeSugar(swift::Demangle::Demangler &dem,
   });
 }
 
+swift::Demangle::ManglingErrorOr<std::string>
+TypeSystemSwiftTypeRef::TransformModuleName(
+    llvm::StringRef mangled_name,
+    const llvm::StringMap<llvm::StringRef> &module_name_map) {
+  swift::Demangle::Demangler dem;
+  auto *node = dem.demangleSymbol(mangled_name);
+  auto *adjusted_node = TypeSystemSwiftTypeRef::Transform(
+      dem, node, [&](swift::Demangle::NodePointer node) {
+        if (node->getKind() == Node::Kind::Module) {
+          auto module_name = node->getText();
+          if (module_name_map.contains(module_name)) {
+            auto real_name = module_name_map.lookup(module_name);
+            auto *adjusted_module_node =
+                dem.createNodeWithAllocatedText(Node::Kind::Module, real_name);
+            return adjusted_module_node;
+          }
+        }
+
+        return node;
+      });
+
+  auto mangling = mangleNode(adjusted_node);
+  return mangling;
+}
+
+NodePointer TypeSystemSwiftTypeRef::TransformTupleElements(
+    NodePointer node, Demangler &dem,
+    std::function<NodePointer(NodePointer)> transformer) {
+  if (!node || node->getKind() != Node::Kind::Type || node->getNumChildren() != 1)
+    return nullptr;
+
+  NodePointer tuple = node->getFirstChild();
+  if (tuple->getKind() != Node::Kind::Tuple || !node->hasChildren())
+    return nullptr;
+
+  node = swift_demangle::CloneNode(dem, node);
+
+  for (size_t i = 0; i < tuple->getNumChildren(); ++i) {
+    NodePointer tuple_element = tuple->getChild(i);
+    assert(tuple_element->getKind() == Node::Kind::TupleElement && tuple->hasChildren());
+    NodePointer child = tuple_element->getFirstChild();
+    NodePointer transformed = transformer(child);
+    tuple->replaceChild(i, transformed);
+  }
+  return node;
+}
+
+NodePointer TypeSystemSwiftTypeRef::TransformModuleName(
+    NodePointer node, Demangler &dem,
+    std::function<NodePointer(NodePointer)> module_transformer) {
+  assert(node->getKind() == Node::Kind::Type);
+  if (node->getKind() != Node::Kind::Type) {
+    LLDB_LOGF(GetLog(LLDBLog::Types),
+              "[TransformModuleName] Unexpected node kind %hu",
+              static_cast<uint16_t>(node->getKind()));
+    return node;
+  }
+
+  bool failed = false;
+  std::function<NodePointer(NodePointer)> recurse_to_module_node =
+      [&](NodePointer node) -> NodePointer {
+    if (failed)
+      return nullptr;
+    // If the node has children this is not the module node yet, so recursively
+    // examine the first child.
+    if (node->hasChildren()) {
+      NodePointer transformed = recurse_to_module_node(node->getFirstChild());
```

Comment:
Because a nominal type's module will the at position 0 at some depth.

For example, a simple struct:
```
struct A {}
```

Will have mangling tree:
```
kind=Global
  kind=TypeMangling
    kind=Type
      kind=Structure
        kind=Module, text="Module"
        kind=Identifier, text="A"
```

A bound generic will still have the module at position 0, but deeper in the tree:

```
kind=Global
  kind=TypeMangling
    kind=Type
      kind=BoundGenericStructure
        kind=Type
          kind=Structure
            kind=Module, text="Swift"
            kind=Identifier, text="Array"
        // TypeList subtree here
```

The module node can be at some arbitrary depth since you can have arbitrarily nested types.